File: src/main/java/fi/metatavu/oioi/cm/rest/translate/AbstractTranslator.java
package fi.metatavu.oioi.cm.rest.translate;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Abstract translator class
 * 
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
public abstract class AbstractTranslator<E, R> {
  
  public abstract R translate(E entity);

  /**
   * Translates list of entities
   * 
   * @param entities list of entities to translate
   * @return List of translated entities
   */
  public List<R> translate(List<E> entities) {
    return entities.stream().map(this::translate).collect(Collectors.toList());
  }

}

File: src/main/kotlin/fi/metatavu/oioi/cm/rest/translate/AbstractTranslator.kt
package fi.metatavu.oioi.cm.rest.translate

/**
 * Abstract translator class
 *
 * @author Jari Nykänen
 */
abstract class AbstractTranslator<E, R> {

    /**
     * Translates single entity
     *
     * @param entity entity to translate
     * @return translated entity
     */
    abstract fun translate(entity: E): R

    /**
     * Translates list of entities
     *
     * @param entities list of entities to translate
     * @return List of translated entities
     */
    open fun translate(entities: List<E>): List<R> {
        return entities.mapNotNull(this::translate)
    }

}